# Tasks 



In [2]:
# Imports
import pandas as pd
import numpy as np
import sklearn as skl

### Task 1: Source the Data Set


In [13]:
# import the Iris data set from the sklearn.datasets module.
from sklearn.datasets import load_iris

# load the iris dataset
data = skl.datasets.load_iris()


#### 'Explain, in your own words, what the `load_iris()` function returns.'

The Iris data set is a data set which collects a number of samples from different species of irises, and is often used in statistics and machine learning, as mentioned [**here**](https://archive.ics.uci.edu/dataset/53/iris) by a paper published by **UC Irvine**.

Running the `load_iris()` function returns a Bunch object, which contains a data array, and all of the necessary keys to use the data set efficiently. This function will now allow me to explore this dataset, which you can see me do below in **Task 2**. 



### Task 2: Explore the Data Structure

#### 'Print and explain the shape of the data set, the first and last 5 rows of the data, the feature names, and the target classes.'

In [18]:
# print the data shape
print(data['data'].shape)

(150, 4)


The output shows us `(150, 4)`, which is a NumPy array of the data set. 150, being the 50 instances of each Iris class, mentioned previously in the UCI paper, whereas 4 is the number of characteristics measured from each sample. Let's explore further into the data set.

In [9]:
# load the dataset in a dataframe
df = pd.read_csv("https://gist.githubusercontent.com/curran/a08a1080b88344b0c8a7/raw/0e7a9b0a5d22642a06d3d5b9bcbad9890c8ee534/iris.csv")

In [10]:
# show dataframe
df

sepal_length  sepal_width  petal_length  petal_width    species
0             5.1          3.5           1.4          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2     setosa
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2     setosa
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146           6.3          2.5           5.0          1.9  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          1.8  virginica

[150 rows x 5 columns]

In [14]:
# show keys
data.keys()


dict_keys(['data', 'target', 'frame', 'target_names', 'DESCR', 'feature_names', 'filename', 'data_module'])

In [21]:
# show features
data['data']


array([[5.1, 3.5, 1.4, 0.2],
       [4.9, 3. , 1.4, 0.2],
       [4.7, 3.2, 1.3, 0.2],
       [4.6, 3.1, 1.5, 0.2],
       [5. , 3.6, 1.4, 0.2],
       [5.4, 3.9, 1.7, 0.4],
       [4.6, 3.4, 1.4, 0.3],
       [5. , 3.4, 1.5, 0.2],
       [4.4, 2.9, 1.4, 0.2],
       [4.9, 3.1, 1.5, 0.1],
       [5.4, 3.7, 1.5, 0.2],
       [4.8, 3.4, 1.6, 0.2],
       [4.8, 3. , 1.4, 0.1],
       [4.3, 3. , 1.1, 0.1],
       [5.8, 4. , 1.2, 0.2],
       [5.7, 4.4, 1.5, 0.4],
       [5.4, 3.9, 1.3, 0.4],
       [5.1, 3.5, 1.4, 0.3],
       [5.7, 3.8, 1.7, 0.3],
       [5.1, 3.8, 1.5, 0.3],
       [5.4, 3.4, 1.7, 0.2],
       [5.1, 3.7, 1.5, 0.4],
       [4.6, 3.6, 1. , 0.2],
       [5.1, 3.3, 1.7, 0.5],
       [4.8, 3.4, 1.9, 0.2],
       [5. , 3. , 1.6, 0.2],
       [5. , 3.4, 1.6, 0.4],
       [5.2, 3.5, 1.5, 0.2],
       [5.2, 3.4, 1.4, 0.2],
       [4.7, 3.2, 1.6, 0.2],
       [4.8, 3.1, 1.6, 0.2],
       [5.4, 3.4, 1.5, 0.4],
       [5.2, 4.1, 1.5, 0.1],
       [5.5, 4.2, 1.4, 0.2],
       [4.9, 3

In [22]:
# show feature names
data['feature_names']

['sepal length (cm)',
 'sepal width (cm)',
 'petal length (cm)',
 'petal width (cm)']

In [24]:
# show target names
data['target_names']

array(['setosa', 'versicolor', 'virginica'], dtype='<U10')

In [25]:
# create dataframe
df.describe()


sepal_length  sepal_width  petal_length  petal_width
count    150.000000   150.000000    150.000000   150.000000
mean       5.843333     3.054000      3.758667     1.198667
std        0.828066     0.433594      1.764420     0.763161
min        4.300000     2.000000      1.000000     0.100000
25%        5.100000     2.800000      1.600000     0.300000
50%        5.800000     3.000000      4.350000     1.300000
75%        6.400000     3.300000      5.100000     1.800000
max        7.900000     4.400000      6.900000     2.500000

### Task 3: Summarize the Data

For each feature in the dataset, calculate and display:

- mean
- minimum
- maximum
- standard deviation
- median


### Task 4: Visualize Features

Plot histograms for each feature using `matplotlib`.

Add appropriate titles and axis labels.

### Task 5: Investigate Relationships

Choose any two features from the data set and create a scatter plot of them.

Color-code the three different classes of the scatter plot points.

### Task 6: Analyze Relationship

Use `numpy.polyfit` to add a regression line to the scatter plot from Task 5.

### Task 7: Analyze Class Distributions

Create box-plots of the petal lengths for each of the three classes.

### Task 8: Compute Correlations

Calculate the correlation coefficients between the features.

Display the results as a heatmap using `matplotlib`.

### Task 9: Fit a Simple Linear Regression

For your two features in Task 5, calculate the coefficient of determination $R^2$.

Re-create the plot from Task 6 and annotate it with the $R^2$ value.

### Task 10: Too Many Features

Use `seaborn` to create a `pairplot` of the data set.

Explain, in your own words, what the `pairplot` depicts.